src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -443,30 +451,55 @@ internal TexlBinding ApplyBindingInternal()
                     }
                 }
 
-                if (this.ReturnType != null && this.ReturnType != FormulaType.Blank && this._expectedReturnType != null)
+                if (this.ReturnType != null && this.ReturnType != FormulaType.Blank)
                 {
-                    bool notCoerceToType = false;
-                    if (_allowCoerceToType)
+                    if (this._expectedReturnType != null)
                     {
-                        switch (this._expectedReturnType)
+                        bool notCoerceToType = false;
+                        if (_allowCoerceToType)
                         {
-                            case StringType:                                
-                                notCoerceToType = !StringValue.AllowedListConvertToString.Contains(this.ReturnType);
-                                break;
-                            case NumberType:
-                                notCoerceToType = !NumberValue.AllowedListConvertToNumber.Contains(this.ReturnType);
-                                break;
-                            case DecimalType:
-                                notCoerceToType = !DecimalValue.AllowedListConvertToDecimal.Contains(this.ReturnType);
-                                break;
-                            default:
-                                throw new NotImplementedException($"Setting ExpectedReturnType to {_expectedReturnType.GetType().FullName} is not implemented");
+                            switch (this._expectedReturnType)
+                            {
+                                case StringType:
+                                    notCoerceToType = !StringValue.AllowedListConvertToString.Contains(this.ReturnType);
+                                    break;
+                                case NumberType:
+                                    notCoerceToType = !NumberValue.AllowedListConvertToNumber.Contains(this.ReturnType);
+                                    break;
+                                case DecimalType:
+                                    notCoerceToType = !DecimalValue.AllowedListConvertToDecimal.Contains(this.ReturnType);
+                                    break;
+                                default:
+                                    throw new NotImplementedException($"Setting ExpectedReturnType to {_expectedReturnType.GetType().FullName} is not implemented");
+                            }
+                        }

```

Comment:
Can this be moved into SetExpectedReturnValue?  Since that now handles arrays of types, and this is just selecting the right array. 

And then we can drop _allowCoerceToType altogether

Reply:
Updated.